File: app/src/main/java/xyz/mordorx/sicmu/ShuffleMode.java
package xyz.mordorx.sicmu;

import android.view.View;

import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;

public enum ShuffleMode {
    SEQUENTIAL(0),
    RANDOM(1),
    RADIO(2); /// Radio is basically random selection (above) plus random song start time when playing the first song

    public final int num;
    ShuffleMode(int num) {
        this.num = num;
    }

    /// This returns the next shuffle mode in sequence, and loops when reaching the last one.
    public ShuffleMode next() {
        var modes = ShuffleMode.values().length;
        var next = (this.num + 1) % modes;
        return ShuffleMode.valueOf(next);
    }

    /// Whether to start playback mid-song. This applies to RADIO.
    public boolean startMidSong() {
        return (this == RADIO);
    }

    public boolean randomSongOrder() {
        return (this == RANDOM || this == RADIO);
    }

    public void showExplainSnackbar(View v) {
        var ctx = v.getContext();
        var txt = "";
        switch (this) {
            case SEQUENTIAL:
                txt = ctx.getString(R.string.settings_shuffle_explainer_sequential);
                break;
            case RANDOM:
                txt = ctx.getString(R.string.settings_shuffle_explainer_random);
                break;
            case RADIO:
                txt = ctx.getString(R.string.settings_shuffle_explainer_radio);
                break;
        }

        Snackbar.make(v, txt, BaseTransientBottomBar.LENGTH_SHORT).show();
    }

    public static ShuffleMode valueOf(int a) {
        for (ShuffleMode m : ShuffleMode.values()) {
            if (m.num == a) {
                return m;
            }
        }
        throw new IllegalArgumentException("Shuffle mode is invalid");
    }
}

File: app/src/main/java/xyz/mordorx/sicmu/media/ShuffleMode.kt
package xyz.mordorx.sicmu.media

import android.view.View
import com.google.android.material.snackbar.BaseTransientBottomBar
import com.google.android.material.snackbar.Snackbar
import xyz.mordorx.sicmu.R

enum class ShuffleMode(
    /** Radio is basically random selection (above) plus random song start time when playing the first song */
    val num: Int
) {
    SEQUENTIAL(0),
    RANDOM(1),
    RADIO(2);

    /** This returns the next shuffle mode in sequence, and loops when reaching the last one. */
    fun next(): ShuffleMode {
        val modes = entries.size
        val next = (this.num + 1) % modes
        return valueOf(next)
    }

    /** Whether to start playback mid-song. This applies to RADIO. */
    fun startMidSong(): Boolean {
        return (this == ShuffleMode.RADIO)
    }

    fun randomSongOrder(): Boolean {
        return (this == ShuffleMode.RANDOM || this == ShuffleMode.RADIO)
    }

    fun showExplainSnackbar(v: View) {
        val ctx = v.getContext()
        var txt = ""
        when (this) {
            ShuffleMode.SEQUENTIAL -> txt =
                ctx.getString(R.string.settings_shuffle_explainer_sequential)

            ShuffleMode.RANDOM -> txt = ctx.getString(R.string.settings_shuffle_explainer_random)
            ShuffleMode.RADIO -> txt = ctx.getString(R.string.settings_shuffle_explainer_radio)
        }

        Snackbar.make(v, txt, BaseTransientBottomBar.LENGTH_SHORT).show()
    }

    companion object {
        fun valueOf(a: Int): ShuffleMode {
            for (m in entries) {
                if (m.num == a) {
                    return m
                }
            }
            throw IllegalArgumentException("Shuffle mode is invalid")
        }
    }
}
